src/itdelatrisu/opsu/beatmap/BeatmapParser.java

```diff
@@ -732,12 +734,21 @@ public static void parseHitObjects(Beatmap beatmap) {
 					hitObject.setComboIndex(comboIndex);
 					hitObject.setComboNumber(comboNumber++);
 
-					beatmap.objects[objectIndex++] = hitObject;
+					objects.add(hitObject);
+					objectIndex++;
 				} catch (Exception e) {
 					Log.warn(String.format("Failed to read hit object '%s' for beatmap '%s'.",
 							line, beatmap.toString()), e);
 				}
 			}
+			objects.trimToSize();
+			Collections.sort(objects, new Comparator<HitObject>() {
```

Comment:
Is sorting really necessary? Does the osu! editor actually ever put the hit objects out of order (in time)?

Anyway, assuming there are incorrectly-ordered hit objects, sorting it like this would corrupt the combo index/number for the whole map.

Reply:
I don't think you should use Collections in this case, as there would be instances they would derail out of order map-wise

Reply:
The sorting is for "Crashpire" to load (and compute its star rating correctly), as according to its beatmap code:

```
[HitObjects]
425,152,250138,70,0,B|35556:20000|35556:20000|-16384:442,1,97280,4|0,0:0|0:0,0:0:0:0:
301,92,251370,54,0,L|262144:262144,1,194560
261,231,19692,6,0,P|306:209|329:150,2,116.5,0|0|2,0:0|0:0|0:0,0:0:0:0:
```

The first two objects are found at the end of the beatmap, and the "true" first note is actually the third (based on the timestamp in parameter two).
This _is_ unique to Aspire maps (as usually notes are placed in order of their time)

As for the Collections, would synchronizing prevent said instances?
I did only use them to create the hit object array.